osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,252 @@
+        yield(TableRowHolder(new DynamicTableRow(std::move(row))));
+      }
+    }
+
+    ret = EvtNext(
+        queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+  }
+  if (ERROR_NO_MORE_ITEMS != GetLastError()) {
+    // No need to close the handler after error; The query
+    // EvtClose will also close all the event handler
+    VLOG(1) << "EvtNext failed with error " << GetLastError();
+  }
+}
+
+void genXfilterFromConstraints(QueryContext& context, std::string& xfilter) {
+  std::vector<std::string> xfilterList;
+
+  auto eids = context.constraints["eventid"].getAll(EQUALS);
+  if (!eids.empty()) {
+    xfilterList.emplace_back(
+        "(EventID=" + osquery::join(eids, ") or (EventID=") + ")");
+  }
+
+  auto pids = context.constraints["pid"].getAll(EQUALS);
+  if (!pids.empty()) {
+    xfilterList.emplace_back(
+        "(Execution[@ProcessID=" +
+        osquery::join(pids, "]) or (Execution[@ProcessID=") + "])");
+  }
+
+  auto times = context.constraints["time_range"].getAll(EQUALS);
+  auto timestamps = context.constraints["timestamp"].getAll(EQUALS);
+  if (!times.empty()) {
+    auto datetime = *times.begin();
+    auto time_vec = osquery::split(datetime, ";");
+
+    if (time_vec.size() == 1) {
+      auto _start = time_vec.front();
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start + "']");
+    } else if (time_vec.size() == 2) {
+      auto _start = time_vec.front();
+      auto _end = time_vec.at(1);
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start +
+                               "' and @SystemTime&lt;='" + _end + "']");
+    }
+  } else if (!timestamps.empty()) {
+    auto time_diff = *timestamps.begin();
+    xfilterList.emplace_back(
+        "TimeCreated[timediff(@SystemTime) &lt;= " + time_diff + "]");
+  }
+
+  xfilter = xfilterList.empty()
+                ? "*"
+                : "*[System[" + osquery::join(xfilterList, " and ") + "]]";
+}
+
+void genWindowsEventLog(RowYield& yield, QueryContext& context) {
+  std::set<std::pair<std::string, std::string>> xpath_set;
+
+  // Check if the `xpath` constraint is available. The `xpath` has
+  // priority over `channel` and it will lookup and filter the events
+  // based on `xpath` if passed.
+  if (context.hasConstraint("xpath", EQUALS)) {
+    auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+    auto xpath = *xpaths.begin();
+    pt::ptree propTree;
+    std::stringstream ss;
+    ss << xpath;
+    pt::read_xml(ss, propTree);
+    auto channel = propTree.get("QueryList.Query.Select.<xmlattr>.Path", "");
+    if (!channel.empty()) {
+      xpath_set.insert(std::make_pair(channel, xpath));
+    } else {
+      LOG(WARNING) << "Invalid xpath format - " << xpath;
+    }
+
+  } else if (context.hasConstraint("channel", EQUALS)) {
+    auto channels = context.constraints["channel"].getAll(EQUALS);
+    std::string xfilter("");
+    genXfilterFromConstraints(context, xfilter);
```

Comment:
Does the event API require that channel is specified in order to provide xfilters? If not, can we move this out of the `channel` constraint?

Reply:
The interface is designed where you can make a query either with `channel` or `xpath` and `xpath` will have priority over the `channel`.  The generation of `xfilter` string is only needed if the query is made with channel constraint. That's why `genXfilterFromConstraints` gets called inside it.